### PR TITLE
Raymarcher that can compute absorption

### PR DESCRIPTION
**Pull request by Biart95**
This PR contains basic raymarcher that is integrated to the PathTracer code, which can render absorption-only participating media.

### CODE REVIEW

**Comment by Biart95** [comment]
I also changed DirectLightingIntegrator so that it can be used both for phase function and BSDF. This might be handy while implementing the first version of single scattering. (First distance samplng - then lighting integration just like for BSDF)

File: src/appleseed/renderer/kernel/lighting/pathtracer.h [hunk 1]
```diff
@@ -452,140 +464,185 @@ size_t PathTracer<PathVisitor, Adjoint>::trace(
...
-        if (vertex.m_bsdf != 0 && crossing_interface)
+        if (crossing_interface)
         {
-            // Refracted ray: inherit the medium list of the parent ray and add/remove the current medium.
+            // Ray goes under the surface:
+            // inherit the medium list of the parent ray and add/remove the current medium.
             if (entering)
             {
-                const float ior =
+                const float ior = !vertex.m_bsdf ? 1.0f :
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/pathtracer.h]: We try to use boolean tests (`!x`) for boolean expressions only (i.e. when `x` is of type `bool`) and prefer explicit tests against 0 for other types. Here: `vertex.m_bsdf == nullptr`

File: src/appleseed/renderer/kernel/lighting/materialsamplers.h [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    virtual bool contributes_to_light_sampling() const = 0;
+
+    virtual const ShadingPoint& trace(
+        const ShadingContext&        shading_context,
+        const foundation::Vector3f&  direction,
+        float&                       transmission) const = 0;
+
+    virtual float trace_between(
+        const ShadingContext&        shading_context,
+        const foundation::Vector3d&  target_position
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.h]: Move closing parenthesis to previous line (do the same change in other places).

File: src/appleseed/renderer/kernel/lighting/materialsamplers.h [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        ) const = 0;
+
+    virtual bool sample(
+        SamplingContext&               sampling_context,
+        const foundation::Dual3d&      outgoing,
+        foundation::Dual3f&            incoming,
+        Spectrum&          value,
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.h]: Align arguments.

File: src/appleseed/renderer/kernel/lighting/materialsamplers.h [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+// Standard headers.
+#include <cstddef>
+
+// Forward declarations.
+namespace renderer  { class BSDF; }
+namespace renderer  { class ShadingContext; }
+namespace renderer  { class ShadingPoint; }
+
+//
+// This file contains wrappers over samplable functions, such as BSDF and PhaseFunction,
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.h]: `,` -> `.`

File: src/appleseed/renderer/kernel/lighting/materialsamplers.h [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+// These wrappers provide a uniform interface for sampling and evaluating these functions
+// in a certain point, which can be used during lighting integration.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.h]: *at* a certain point

File: src/appleseed/renderer/kernel/lighting/materialsamplers.h [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    virtual const ShadingPoint& trace(
+        const ShadingContext&          shading_context,
+        const foundation::Vector3f&    direction,
+        float&                         transmission) const override;
+
+    virtual bool cull_incoming_direction(
+        const foundation::Vector3d& incoming) const override;
+
+private:
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.h]: Indent with two spaces.

File: src/appleseed/renderer/kernel/lighting/materialsamplers.cpp [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+bool PhaseFunctionSampler::sample(
+    SamplingContext&   sampling_context,
+    const Dual3d&      outgoing,
+    Dual3f&            incoming,
+    Spectrum&          value,
+    float&             pdf) const
+{
+    Vector3f incoming_direction;
+
+    // Sample the Phase Function.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.cpp]: This comment is probably redundant.

File: src/appleseed/renderer/kernel/lighting/materialsamplers.cpp [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+float PhaseFunctionSampler::evaluate(
+    int              light_sampling_modes,
+    const Vector3f&  outgoing,
+    const Vector3f&  incoming,
+    Spectrum&        value
+    ) const
+{
+    // Evaluate the Phase Function.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.cpp]: This comment is probably redundant.

File: src/appleseed/renderer/kernel/lighting/materialsamplers.cpp [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    Spectrum transmission;
+    m_phasefunction.evaluate_transmission(
+        m_volume_ray, m_phasefunction_data, m_distance, transmission);
+    value *= transmission;
+
+    return pdf;
+}
+
+float PhaseFunctionSampler::trace_between(
+    const ShadingContext& shading_context,
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.cpp]: Align arguments.

File: src/appleseed/renderer/kernel/lighting/materialsamplers.cpp [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        Vector3d(direction),
+        m_volume_ray.m_time,
+        VisibilityFlags::ShadowRay,
+        m_volume_ray.m_depth + 1,
+        transmission);
+}
+
+bool PhaseFunctionSampler::cull_incoming_direction(const Vector3d& incoming) const
+{
+    // No culling for phase function
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.cpp]: Terminate full line comment with a `.`

File: src/appleseed/renderer/kernel/lighting/materialsamplers.cpp [hunk 11]
```diff
@@ -0,0 +1,272 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/materialsamplers.cpp]: Remove empty line.

File: src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp [hunk 12]
```diff
@@ -206,8 +198,9 @@ void DirectLightingIntegrator::compute_outgoing_radiance_light_sampling_low_vari
     if (!m_light_sampler.has_lights_or_emitting_triangles())
         return;
 
-    // There cannot be any contribution for purely specular BSDFs.
-    if (m_bsdf.is_purely_specular())
+    // Check if PDF of the sampler is Dirac delta and therefore it
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp]: Remove *it*.

File: src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp [hunk 13]
```diff
@@ -698,57 +676,46 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
...
         light_value);
 
     // Compute the incoming direction in world space.
     const Vector3d incoming = -emission_direction;
 
-    // Cull light samples behind the shading surface if the BSDF is either reflective or transmissive,
-    // but not both.
-    if (m_bsdf.get_type() != BSDF::AllBSDFTypes)
+    if (m_material_sampler.cull_incoming_direction(incoming))
     {
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp]: I suggest omitting the braces for single-line bodies:

    if (m_material_sampler.cull_incoming_direction(incoming))
        return;

File: src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp [hunk 14]
```diff
@@ -617,29 +601,23 @@ void DirectLightingIntegrator::add_emitting_triangle_sample_contribution(
...
-    // Evaluate the BSDF.
-    Spectrum bsdf_value;
-    const float bsdf_prob =
-        m_bsdf.evaluate(
-            m_bsdf_data,
-            false,              // not adjoint
-            true,               // multiply by |cos(incoming, normal)|
-            Vector3f(m_geometric_normal),
-            Basis3f(m_shading_basis),
+    // Evaluate the BSDF (or phase function scattering).
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp]: Remove *scattering*?

File: src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp [hunk 15]
```diff
@@ -698,57 +676,46 @@ void DirectLightingIntegrator::add_non_physical_light_sample_contribution(
...
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/directlightingintegrator.cpp]: Remove *scattering*?

**Comment by dictoon** [comment]
Thank you!

**Pull request merged**